src/msg/async/AsyncMessenger.cc

```diff
@@ -936,7 +936,15 @@ int AsyncMessenger::reap_dead()
   while (!deleted_conns.empty()) {
     auto it = deleted_conns.begin();
     AsyncConnectionRef p = *it;
-    ldout(cct, 5) << __func__ << " delete " << p << dendl;
+    ldout(cct, 5) << __func__ << " delete " << p
+		  << " (priv " << p->priv << ")" << dendl;
+    if (p->priv) {
+      // there may be a priv <-> session ref cycle; break it by sequencing
+      // clearing the priv pointer *after* anything else that is already in
+      // the dispatch queue.
+      dispatch_queue.queue_reap(p.get());
```

Comment:
please note, we assume that someone is holding a strong reference of `AsyncConnection*` *after* it is removed from `accepting_conns` and `deleted_conns`, and it's very likely `p->priv`. but i don't think we can take it as a fact.

so i'd suggest put something like
```c++
 dispatch_queue.queue_reap(QueueItem(D_CONN_REAP, p.get());
```
and update `queue_reap()` 's signature accordingly. so we can ensure that Connection is not destroyed before we deference it when handling a `D_CONN_REAP` item.

i understand that this change makes `queue_reap()` look weird in its `queue_*()` cousins. but it is indeed different in the sense that it is the only queue method that could be post a connection to the DispatchQueue  after `accepting_conns.erase()` and `deleted_conns.erase()` erase it.

Reply:
The QueueItem member is ConnectionRef, so we're taking a reference in queue_reap(), *both* with the current code and with your more explicit form.  Is the confusing part that the arg is Connection* instead of ConnectionRef&?

Reply:
ahh, thought `queue_reap()` was an async call.. should be safe then.